xheap.py

```diff
@@ -114,31 +114,22 @@ def push(self, item):
         _siftdown(self, 0, len(self)-1)
 
     def pop(self, index=None):
+        index = index or 0
+        if index < 0:
+            index += len(self)
+        if index < 0:
+            raise IndexError
         return_item = self._pop(index)
         del self._indexes[return_item]
         return return_item
 
     def _pop(self, index):
-        if index:
-            last_item = super(Heap, self).pop()
-            if index == len(self):
-                return_value = last_item
-            else:
-                return_value = self[index]
-                self[index] = last_item
-                if self[(index - 1) >> 1] < last_item:
```

Comment:
I understand this for `_siftdown`. But `_siftup` does not check at all. It pulls ALL items up without breaking out as it moves children not the new item:

```
def _siftup(heap, pos):
    endpos = len(heap)
    startpos = pos
    newitem = heap[pos]
    # Bubble up the smaller child until hitting a leaf.
    childpos = 2*pos + 1    # leftmost child position
    while childpos < endpos:
        # Set childpos to index of smaller child.
        rightpos = childpos + 1
        if rightpos < endpos and not heap[childpos] < heap[rightpos]:
            childpos = rightpos
        # Move the smaller child up.
        heap[pos] = heap[childpos]
        pos = childpos
        childpos = 2*pos + 1
    # The leaf at pos is empty now.  Put newitem there, and bubble it up
    # to its final resting place (by sifting its parents down).
    heap[pos] = newitem
    _siftdown(heap, startpos, pos)
```

If you say that's not significant (due to some math or so), that's alright. Just asking.